### PR TITLE
DMP-1890-download-annotation-document

### PR DESCRIPTION
**Pull request by JasonPittockCGI**
JIRA https://tools.hmcts.net/jira/browse/DMP-1890

Introduce download document functionality

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x ] No
```

### CODE REVIEW

File: src/main/resources/openapi/annotations.yaml [hunk 1]
```diff
@@ -51,6 +51,77 @@ paths:
               schema:
                 $ref: './problem.yaml'
 
+  /annotations/{annotation_id}/documents/{annotation_document_id}:
```
Comment (cmagowan-hmcts) [inline src/main/resources/openapi/annotations.yaml:54]: When adding a new endpoint, can you add an example request to the postman collection?

File: src/main/resources/openapi/annotations.yaml [hunk 2]
```diff
@@ -51,6 +51,77 @@ paths:
...
+          required: true
+      responses:
+        '200':
+          description: OK
+          content:
+            application/zip:
+              schema:
+                type: string
+                format: binary
+              example:
```
Comment (cmagowan-hmcts) [inline src/main/resources/openapi/annotations.yaml:81]: this example looks like an error?

File: src/main/java/uk/gov/hmcts/darts/annotation/controller/AnnotationController.java [hunk 3]
```diff
@@ -31,4 +36,30 @@ public ResponseEntity<PostAnnotationResponse> annotationsPost(MultipartFile file
         var annotationId = annotationService.process(file, annotation);
         return ResponseEntity.ok(new PostAnnotationResponse(annotationId));
     }
+
+    @Authorisation(
+            bodyAuthorisation = true,
+            contextId = ANY_ENTITY_ID,
```
Comment (cmagowan-hmcts) [inline src/main/java/uk/gov/hmcts/darts/annotation/controller/AnnotationController.java:42]: we should use the annotation contextId here, but its not been merged yet. Tony is introducing it as part of https://github.com/hmcts/darts-api/pull/1032

File: src/main/java/uk/gov/hmcts/darts/annotation/controller/AnnotationController.java [hunk 4]
```diff
@@ -31,4 +36,30 @@ public ResponseEntity<PostAnnotationResponse> annotationsPost(MultipartFile file
...
+            securityRoles = {JUDGE, ADMIN},
+            globalAccessSecurityRoles = {JUDGE, ADMIN})
+    @Override
+    public ResponseEntity<Resource> downloadAnnotation(Integer annotationId, Integer annotationDocumentId) {
+
+        final AnnotationResponseDto annotationResponseDto = annotationService.downloadAnnotationDoc(1, 1);
```
Comment (cmagowan-hmcts) [inline src/main/java/uk/gov/hmcts/darts/annotation/controller/AnnotationController.java:48]: id's are hardcoded

File: src/main/java/uk/gov/hmcts/darts/annotation/service/impl/AnnotationServiceImpl.java [hunk 5]
```diff
@@ -66,6 +79,46 @@ public Integer process(MultipartFile document, Annotation annotation) {
         return annotationEntity.getId();
     }
 
+    @Override
+    public AnnotationResponseDto downloadAnnotationDoc(Integer annotationId, Integer annotationDocumentId) {
+
+        final Optional<ExternalObjectDirectoryEntity> dirEntity = eodRepository.findAnnotationIdAndAnnotationDocumentId(annotationId, annotationDocumentId);
```
Comment (cmagowan-hmcts) [inline src/main/java/uk/gov/hmcts/darts/annotation/service/impl/AnnotationServiceImpl.java:85]: Shoudl probably have the \by\ word in it, i.e. find**By**AnnotationIdAndAnnotationDocumentId

Comment (cmagowan-hmcts) [inline src/main/java/uk/gov/hmcts/darts/annotation/service/impl/AnnotationServiceImpl.java:85]: `dirEntity` is not very descriptive, should probably be something like `annotationDocEodEntityOpt`

File: src/main/java/uk/gov/hmcts/darts/annotation/service/impl/AnnotationServiceImpl.java [hunk 6]
```diff
@@ -66,6 +79,46 @@ public Integer process(MultipartFile document, Annotation annotation) {
...
+        final InputStreamResource stream;
+
+        final ExternalObjectDirectoryEntity externalObjectDirectoryEntity;
+
+        if (dirEntity.isEmpty()) {
+
+            if (authorisationApi.userHasOneOfRoles(List.of(SecurityRoleEnum.JUDGE))) {
+                throw new DartsApiException(INVALID_ANNOTATIONID_OR_ANNOTATION_DOCUMENTID_FOR_JUDGE);
```
Comment (cmagowan-hmcts) [inline src/main/java/uk/gov/hmcts/darts/annotation/service/impl/AnnotationServiceImpl.java:93]: is there a benefit to the judge specific error message?

File: src/main/resources/openapi/annotations.yaml (hunk 1 above)
Comment (JasonPittockCGI) [reply src/main/resources/openapi/annotations.yaml:54]: Sure. Will do.

File: src/main/java/uk/gov/hmcts/darts/annotation/service/impl/AnnotationServiceImpl.java [hunk 7]
```diff
@@ -66,6 +79,46 @@ public Integer process(MultipartFile document, Annotation annotation) {
...
+                throw new DartsApiException(INVALID_ANNOTATIONID_OR_ANNOTATION_DOCUMENTID);
+            }
+
+        }
+
+        externalObjectDirectoryEntity = dirEntity.get();
+
+        try {
+
+            stream = new InputStreamResource(dataManagementApi.getBlobDataFromInboundContainer(externalObjectDirectoryEntity.getExternalLocation()).toStream());
```
Comment (cmagowan-hmcts) [inline src/main/java/uk/gov/hmcts/darts/annotation/service/impl/AnnotationServiceImpl.java:104]: `stream` is quite vague, maybe `blobStream` or something to give a hint as to what it is

File: src/main/java/uk/gov/hmcts/darts/annotation/service/impl/AnnotationServiceImpl.java [hunk 8]
```diff
@@ -66,6 +79,46 @@ public Integer process(MultipartFile document, Annotation annotation) {
...
+
+        } catch (RuntimeException e) {
+            log.error("Failed to download annotation document {}", externalObjectDirectoryEntity.getExternalFileId(), e);
```
Comment (cmagowan-hmcts) [inline src/main/java/uk/gov/hmcts/darts/annotation/service/impl/AnnotationServiceImpl.java:107]: ExternalFileId is normally blank, i think its only used my ARM. a better field to use might be the eod_id

File: src/main/java/uk/gov/hmcts/darts/annotation/service/impl/AnnotationServiceImpl.java [hunk 9]
```diff
@@ -66,6 +79,46 @@ public Integer process(MultipartFile document, Annotation annotation) {
...
+            throw new DartsApiException(FAILED_TO_DOWNLOAD_ANNOTATION_DOCUMENT, e);
+        }
+
+        return AnnotationResponseDto.builder()
+                .resource(stream)
+                .fileName(externalObjectDirectoryEntity.getAnnotationDocumentEntity().getFileName())
+                .externalLocation(externalObjectDirectoryEntity.getExternalLocation())
+                .annotationDocumentId(externalObjectDirectoryEntity.getAnnotationDocumentEntity().getId()).build();
```
Comment (cmagowan-hmcts) [inline src/main/java/uk/gov/hmcts/darts/annotation/service/impl/AnnotationServiceImpl.java:115]: is this not just the annotationDocumentId thats passed in at the top of the method?

File: src/main/java/uk/gov/hmcts/darts/annotation/service/impl/AnnotationServiceImpl.java [hunk 10]
```diff
@@ -66,6 +79,46 @@ public Integer process(MultipartFile document, Annotation annotation) {
         return annotationEntity.getId();
     }
 
+    @Override
+    public AnnotationResponseDto downloadAnnotationDoc(Integer annotationId, Integer annotationDocumentId) {
```
Comment (cmagowan-hmcts) [inline src/main/java/uk/gov/hmcts/darts/annotation/service/impl/AnnotationServiceImpl.java:83]: i don't see any checks that the user downloading the annotation has access to it, it shoudl be somethgin similar to Tonys:-
https://github.com/hmcts/darts-api/pull/1032/files#diff-3cb4a8770477be43d2a61bc42628126ac2d90df6ca290270f9d73f191a6c1ab2R25


or perhaps wait until his is merged and make use of it

File: src/main/java/uk/gov/hmcts/darts/annotation/service/impl/AnnotationServiceImpl.java (hunk 5 above)
Comment (JasonPittockCGI) [reply src/main/java/uk/gov/hmcts/darts/annotation/service/impl/AnnotationServiceImpl.java:85]: Addressed.

File: src/main/java/uk/gov/hmcts/darts/annotation/service/impl/AnnotationServiceImpl.java (hunk 7 above)
Comment (JasonPittockCGI) [reply src/main/java/uk/gov/hmcts/darts/annotation/service/impl/AnnotationServiceImpl.java:104]: Done.

File: src/main/java/uk/gov/hmcts/darts/annotation/service/impl/AnnotationServiceImpl.java (hunk 8 above)
Comment (JasonPittockCGI) [reply src/main/java/uk/gov/hmcts/darts/annotation/service/impl/AnnotationServiceImpl.java:107]: Done

File: src/main/java/uk/gov/hmcts/darts/annotation/service/impl/AnnotationServiceImpl.java (hunk 9 above)
Comment (JasonPittockCGI) [reply src/main/java/uk/gov/hmcts/darts/annotation/service/impl/AnnotationServiceImpl.java:115]: Done

File: src/main/java/uk/gov/hmcts/darts/annotation/controller/AnnotationController.java (hunk 3 above)
Comment (JasonPittockCGI) [reply src/main/java/uk/gov/hmcts/darts/annotation/controller/AnnotationController.java:42]: merged Tony's ticket, and used Annotation context_id

File: src/main/java/uk/gov/hmcts/darts/annotation/service/impl/AnnotationServiceImpl.java (hunk 6 above)
Comment (JasonPittockCGI) [reply src/main/java/uk/gov/hmcts/darts/annotation/service/impl/AnnotationServiceImpl.java:93]: This was how I interpreted the ticket. Do u think it incorrect?

File: src/integrationTest/java/uk/gov/hmcts/darts/annotation/AnnotationControllerTest.java [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+import static uk.gov.hmcts.darts.common.enums.SecurityRoleEnum.JUDGE;
+import static uk.gov.hmcts.darts.testutils.data.AnnotationTestData.minimalAnnotationEntity;
+import static uk.gov.hmcts.darts.testutils.data.HearingTestData.createSomeMinimalHearing;
+import static uk.gov.hmcts.darts.testutils.data.HearingTestData.someMinimalHearing;
+import static uk.gov.hmcts.darts.testutils.data.SecurityGroupTestData.buildGroupForRole;
+import static uk.gov.hmcts.darts.testutils.data.SecurityGroupTestData.buildGroupForRoleAndCourthouse;
+import static uk.gov.hmcts.darts.testutils.data.UserAccountTestData.minimalUserAccount;
+
+@AutoConfigureMockMvc
+@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
```
Comment (tonymort) [inline src/integrationTest/java/uk/gov/hmcts/darts/annotation/AnnotationControllerTest.java:55]: Is DirtiesContext neccessary? Slow the test suite down

File: src/integrationTest/java/uk/gov/hmcts/darts/annotation/AnnotationControllerTest.java [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+                requestBuilder)
+            .andExpect(status().isNotFound())
+            .andExpect(jsonPath("$.type").value("ANNOTATION_104"))
+            .andExpect(jsonPath("$.title").value("invalid annotation id or annotation document id"))
+            .andExpect(jsonPath("$.status").value("404"))
+            .andReturn();
+    }
+
+    @Test
+    @Transactional
```
Comment (tonymort) [inline src/integrationTest/java/uk/gov/hmcts/darts/annotation/AnnotationControllerTest.java:141]: We dont want to introduce more transactional tests.  What is the reason for it?  If not sure I can take a look.

File: src/main/resources/openapi/annotations.yaml [hunk 13]
```diff
@@ -96,3 +163,27 @@ components:
...
+      type: string
+      enum:
+        - "The requested hearing cannot be found"
+        - "The annotation failed to be uploaded"
+        - "user is not authorised to download annotation documents"
+        - "annotation ID or annotation document ID does not exist"
+        - "invalid annotation id or annotation document id"
+        - "invalid annotation id or annotation document id for judge"
+        - "The annotation failed to be downloaded"
+      x-enum-varnames: [ HEARING_NOT_FOUND, FAILED_TO_UPLOAD_ANNOTATION_DOCUMENT, USER_NOT_AUTHORISED_TO_DOWNLOAD, INVALID_ANNOTATIONID_OR_ANNOTATION_DOCUMENTID, INVALID_ANNOTATIONID_OR_ANNOTATION_DOCUMENTID_FOR_JUDGE, FAILED_TO_DOWNLOAD_ANNOTATION_DOCUMENT]
```
Comment (cmagowan-hmcts) [inline src/main/resources/openapi/annotations.yaml:189]: the enum is not aligned
image

File: src/main/java/uk/gov/hmcts/darts/annotation/controller/AnnotationController.java [hunk 14]
```diff
@@ -43,4 +46,30 @@ public ResponseEntity<Void> deleteAnnotation(Integer annotationId) {
         annotationService.delete(annotationId);
         return ResponseEntity.noContent().build();
     }
+
+    @Authorisation(
+            bodyAuthorisation = true,
```
Comment (cmagowan-hmcts) [inline src/main/java/uk/gov/hmcts/darts/annotation/controller/AnnotationController.java:51]: annotationId is in the path, not the body, so `bodyAuthorisation = true,` is not required

File: src/main/java/uk/gov/hmcts/darts/annotation/service/impl/AnnotationServiceImpl.java (hunk 6 above)
Comment (cmagowan-hmcts) [reply src/main/java/uk/gov/hmcts/darts/annotation/service/impl/AnnotationServiceImpl.java:98]: i've double checked with Dave, and there is no difference between Admin and Judge errors. i have updated the ACs to clarify

File: src/integrationTest/java/uk/gov/hmcts/darts/annotation/AnnotationControllerTest.java (hunk 12 above)
Comment (JasonPittockCGI) [reply src/integrationTest/java/uk/gov/hmcts/darts/annotation/AnnotationControllerTest.java:141]: Hi Tony. Detached entity. So ran the test with a transaction. Also had to use dirties context to clear data between  test methods

**Review by cmagowan-hmcts — APPROVED** [review]

**Pull request merged**